client/src/components/PresentedOptions.js

```diff
@@ -0,0 +1,504 @@
+  constructor(props) {
+    super(props);
+
+    const value = props.value || {};
+    this.state = {
+      customOptions: [],
+      selectType: props.selectTypeDefault,
+      selections: {},
+      suggestedOptions: [],
+      suggestedOptionCache: {},
+      loading: false,
+      separatorDelimiter: '',
+      usedSeparators: [],
+      ...value,
+    };
+
+    this.handleCheckboxChange = this.handleCheckboxChange.bind(this);
+    this.handleInputChange = this.handleInputChange.bind(this);
+    this.handleSelectTypeChange = this.handleSelectTypeChange.bind(this);
+    this.handleDelimiterChange = this.handleDelimiterChange.bind(this);
+    this.handleExecuteSeparator = this.handleExecuteSeparator.bind(this);
+  }
+
+  componentDidMount() {
+    // Ensure the suggested options are loaded
+    this.loadSuggestedOptions();
+  }
+
+  componentDidUpdate(prevProps) {
+    // Compare if selected fields have changed
+    const newFields = this.props.selectedFields;
+    const oldFields = prevProps.selectedFields;
+
+    // If the type or length doesn't match then they must be different
+    if (typeof newFields !== typeof oldFields || newFields.length !== oldFields.length) {
+      this.loadSuggestedOptions();
+      return;
+    }
+
+    // Otherwise filter out options that appear in both arrays and assert the remaining count is
+    // zero
+    if (oldFields.filter(old => !newFields.includes(old)).length === 0) {
+      return;
+    }
+
+    this.loadSuggestedOptions();
+  }
+
+  /**
+   * Returns the namespaced form field name for the given field
+   *
+   * @param {string} fieldName
+   * @param {object} props If provided, will be used instead of this.props
+   * @returns {string}
+   */
+  getFieldName(fieldName, props = {}) {
+    const name = props.name || this.props.name;
+    return `${name}-${fieldName}`;
+  }
+
+  /**
+   * Returns the value for the text input from the state
+   *
+   * @returns {string}
+   */
+  getInputValue() {
+    return this.state.customOptions.join('\n');
+  }
+
+  /**
+   * Returns the select type selection, either from the state or from the default value in props
+   *
+   * @returns {string}
+   */
+  getSelectType() {
+    if (typeof this.state.selectType !== 'undefined') {
+      return String(this.state.selectType);
+    }
+    return String(this.props.data.selectTypeDefault);
+  }
+
+  loadSuggestedOptions() {
+    const { selectedFields, data: { endpoint, resource } } = this.props;
+
+    // Clear the state of suggested options...
+    this.setState({
+      suggestedOptions: [],
+      loading: false,
+    });
+
+    // If there's no columns selected we can early return
+    if (!selectedFields.length) {
+      return;
+    }
+
+    let options = [];
+    const { suggestedOptionCache } = this.state;
+    const loadPromises = [];
+
+    // Prep a datastore. Note that this doesn't actually fire any requests yet.
+    const Datastore = CKANApi.loadDatastore(endpoint, resource);
```

Comment:
I think convention would be for this to be lowercase unless it's a React component